xip/xipether.c

```diff
@@ -0,0 +1,449 @@
+	struct xia_xid dst;
+	const char *dev;
+	char strid[2 * XIA_XID_MAX];
+	unsigned char lladdr[MAX_ADDR_LEN];
+	unsigned int oif, addrlen;
+
+	if (argc != 2) {
+		fprintf(stderr, "Wrong number of parameters\n");
+		return usage();
+	}
+
+	if (strcmp(argv[0], "dev")) {
+		fprintf(stderr, "Wrong parameters\n");
+		return usage();
+	}
+
+	dev = argv[1];
+	oif = ll_name_to_index(dev);
+	if (!oif) {
+		fprintf(stderr, "Cannot find device '%s'\n", dev);
+		return -1;
+	}
+	addrlen = ll_index_to_addr(oif, lladdr, sizeof(lladdr));
+	if (!addrlen) {
+		fprintf(stderr, "Cannot find device address '%s'\n", dev);
+		return -1;
+	}
+	if (form_ether_xid(oif, lladdr, XIA_MAX_STRID_SIZE, strid) < 0) {
+		fprintf(stderr, "Cannot form ether XID from given parameters\n");
+		return -1;
+	}
+
+	xrt_get_ppal_id("ether", usage, &dst, strid);
+	return modify_local(&dst, to_add);
+}
+
+static int do_addlocal(int argc, char **argv)
+{
+	return do_local(argc, argv, 1);
+}
+
+static int do_dellocal(int argc, char **argv)
+{
+	return do_local(argc, argv, 0);
+}
+
+static int modify_neigh(struct xia_xid *dst, unsigned char *lladdr,
+	int lladdr_len, unsigned oif, int to_add)
+{
+	struct {
+		struct nlmsghdr 	n;
+		struct rtmsg 		r;
+		char   			buf[1024];
+	} req;
+
+	memset(&req, 0, sizeof(req));
+
+	req.n.nlmsg_len = NLMSG_LENGTH(sizeof(struct rtmsg));
+
+	if (to_add) {
+		/* XXX Does one really needs all these flags? */
+		req.n.nlmsg_flags = NLM_F_REQUEST|NLM_F_CREATE|NLM_F_EXCL;
+		req.n.nlmsg_type = RTM_NEWROUTE;
+		req.r.rtm_scope = RT_SCOPE_LINK;
+	} else {
+		req.n.nlmsg_flags = NLM_F_REQUEST;
+		req.n.nlmsg_type = RTM_DELROUTE;
+		req.r.rtm_scope = RT_SCOPE_NOWHERE;
+	}
+
+	req.r.rtm_family = AF_XIA;
+	req.r.rtm_table = XRTABLE_MAIN_INDEX;
+	req.r.rtm_protocol = RTPROT_BOOT;
+	req.r.rtm_type = RTN_UNICAST;
+
+	req.r.rtm_dst_len = sizeof(*dst);
+	addattr_l(&req.n, sizeof(req), RTA_DST, dst, sizeof(*dst));
+	addattr_l(&req.n, sizeof(req), RTA_LLADDR, lladdr, lladdr_len);
+	addattr32(&req.n, sizeof(req), RTA_OIF, oif);
+
+	if (rtnl_talk(&rth, &req.n, 0, 0, NULL, NULL, NULL) < 0)
+		exit(2);
+	return 0;
+}
+
+static int do_Xneigh_common(int argc, char **argv, int to_add)
+{
+	struct xia_xid dst;
+	char *str_lladdr;
+	char strid[XIA_MAX_STRID_SIZE];
+	unsigned char lladdr[MAX_ADDR_LEN];
+	int lladdr_len;
+	const char *dev;
+	unsigned int oif;
+
+	if (argc != 4) {
+		fprintf(stderr, "Wrong number of parameters\n");
+		return usage();
+	}
+	if (strcmp(argv[0], "lladdr") || strcmp(argv[2], "dev")) {
+		fprintf(stderr, "Wrong parameters\n");
+		return usage();
+	}
+
+	str_lladdr = argv[1];
+	lladdr_len = lladdr_pton(str_lladdr, lladdr, sizeof(lladdr));
+	if (lladdr_len < 0) {
+		fprintf(stderr, "Invalid link layer address: '%s'\n",
+			str_lladdr);
+		return lladdr_len;
+	}
+
+	dev = argv[3];
+	oif = ll_name_to_index(dev);
+	if (!oif) {
+		fprintf(stderr, "Cannot find device '%s'\n", dev);
+		return -1;
+	}
+	if (form_ether_xid(oif, lladdr, XIA_MAX_STRID_SIZE, strid) < 0) {
```

Comment:
Variable `lladdr` has a variable length, but we are not passing its length to `form_ether_xid()`.

Modifying `form_ether_xid()` to deal with the variable size of variable `lladdr` is more work than what we need since we don't expect Ethernet addresses to change size anytime soon. Thus, just modify the line `if (lladdr_len < 0) {` above to `if (lladdr_len != ETHER_ADDR_SIZE) {`